### PR TITLE
Perform dual-fetch when loading the first page of the Processing tab

### LINKED ISSUE #1962: Orders: Refreshing the Processing tab doesn't sync order status with remote changes
**Describe the bug**

If an order with `Processing` status has its status updated remotely (e.g. the order is refunded in WP Admin), it isn't immediately removed from the Processing tab in the app. Pulling to refresh the Processing tab also doesn't update the order status or remove that order from the list.

The only way I found to update the order to the correct status is to switch to the All Orders tab and pull to refresh, or open the order detail and pull to refresh it there.

**To Reproduce**
1. Go to the Orders tab.
2. Select the Processing tab and identify a recent order in the list.
3. In WP Admin web, open that order and fully refund it (confirm the order status is changed to `Refunded`).
4. Open the app again and notice that the order is still in the list on the Processing tab.
5. Pull to refresh the list on the Processing tab. Notice the order is still in the list.
6. Switch to the All Orders tab and pull to refresh. Notice that the order status updates and the order is now removed from the list on the Processing tab.

**Mobile Environment**
Please include:
- Device: iPhone XS
- iOS version: 13.3.1
- WooCommerce iOS version: 3.8.0.0 TestFlight beta

### PR DESCRIPTION
**Pull request by shiki**
Closes #1962

😱 This is **almost 2K lines**. Mainly because I added a JSON file (1.2K lines) for testing. Also, there are lots of comments. And the tests are 200+. 

## Findings

The reason for the bug is simply because, on the Processing tab, we only fetch using `GET /orders?status=processing`. If a _processing_ order was downloaded and was changed to a different status on the web, it will **no longer be returned on** that endpoint query result. So, the app has no chance to update it until we fetch the All Orders list.

This is a pre-existing issue and is just more visible because the Processing tab now exists. 

## Solution

We've discussed a few solutions to this. We considered using the ListStore concept used in WCAndroid. After discussing with everyone, we decided against that because: 

1. It's going to take a while to implement.
2. We're pushing for a better (mobile-friendly) sync endpoint (p7bje6-1IQ-p2). We will invest significant development time on that instead.

The "fix" is to perform these when pulling-to-refresh: 

1. Perform a dual-fetch (`GET ?status=processing` and `GET ?status=any`) concurrently.
2. Delete all the orders before saving the newly fetched ones. 

### Why a Dual Fetch

We could just delete all the orders and perform a single fetch of `GET /orders?status=processed`. But this could result in the user viewing an incomplete All Orders tab until we queried again. It's a jarring experience. Consider this scenario:

1. Navigate to Processing and All Orders tab to load their content.
2. Pull to refresh on the Processing tab.
3. Navigate back to the All Orders tab.

On Step 3, since we deleted all the orders and only downloaded new processing orders, the All Orders tab will initially contain just those orders.

The dual fetch is a "hack" to hide the scenario.

### Deleting all the Orders

This sync strategy doesn't fix all our problems. I'm sure there are other scenarios that can cause stale data. In fact, synchronization by `pageNumber` almost always has bugs because orders could be on a different (previous) page when you load the next page. That order would then end up not getting loaded at all.

Deleting all the orders during a pull to refresh is a "confidence button" for our users.  When they use this, we are guaranteeing them that they will be viewing an up to date list.

### Spec

This is how sync behaves in different scenarios.

| Action           | Current Tab | Delete All | GET ?status=processing | GET ?status=any |
|------------------|-------------|------------|------------------------|-----------------|
| Pull-to-refresh  | Processing  | y          | y                      | y               |
| `viewWillAppear` | Processing  | .          | y                      | y               |
| Load next page   | Processing  | .          | y                      | .               |
| Pull-to-refresh  | All Orders  | y          | .                      | y               |
| `viewWillAppear` | All Orders  | .          | .                      | y               |
| Load next page   | All Orders  | .          | .                      | y               |

### [`OrderStore.fetchFilteredAndAllOrders`](https://github.com/woocommerce/woocommerce-ios/blob/f0e5093e2965309e61713ca5f38c5b7bb3300c94/Yosemite/Yosemite/Stores/OrderStore.swift#L98)

I added this new method which performs the dual fetch and the deletion of the orders. 

https://github.com/woocommerce/woocommerce-ios/blob/f0e5093e2965309e61713ca5f38c5b7bb3300c94/Yosemite/Yosemite/Stores/OrderStore.swift#L86-L102

### [`OrdersViewModel.synchronizationAction`](https://github.com/woocommerce/woocommerce-ios/blob/f0e5093e2965309e61713ca5f38c5b7bb3300c94/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L86-L91)

That `OrderStore` method's `OrderAction` is created inside the `OrdersViewModel.synchronizationAction` method. That method is now used by `OrdersViewController` when it performs a sync:

https://github.com/woocommerce/woocommerce-ios/blob/f0e5093e2965309e61713ca5f38c5b7bb3300c94/WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift#L303

I created this new `ViewModel` so I can easily [create tests](https://github.com/woocommerce/woocommerce-ios/blob/issue/1962-orders-sync/WooCommerce/WooCommerceTests/ViewRelated/Orders/OrdersViewModelTests.swift) for it since this is one of the integral parts of this solution. 

### `SyncCoordinator` `reason` arguments

I needed to be able to distinguish in [`OrdersViewController.sync`](https://github.com/woocommerce/woocommerce-ios/blob/f5c17ebfc308162fb0f52cc3729be0e1dcb3cf0c/WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift#L287) whether the sync request is happening because of a pull-to-refresh or just a normal subsequent page load request. To facilitate this, I added `reason` arguments to the `SyncCoordinator`. 

This `reason` argument is only passed around so that when `SyncCoordinator` calls the `delegate.sync`, that `delegate` can decide how to perform the network request.

image

We use that `reason` in `OrdersViewModel.synchronizationAction` to determine if the sync request was because of a pull-to-refresh: 

https://github.com/woocommerce/woocommerce-ios/blob/f0e5093e2965309e61713ca5f38c5b7bb3300c94/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L93

### Changed `synchronizeFirstPage()` to `resynchronize()`

Wherever `synchronizeFirstPage()` is called in `OrdersViewController`, I changed them to `resynchronize()` instead. This is to reset the `SyncCoordinator`'s internal state and ensure that subsequent pages will be loaded. 

### `MockupNetwork` bug fix

In 06028eb02065e62362928a6a770da3c99c55cbe2, I fixed a bug in `MockupNetwork`'s dequeuing routine. I needed to fix this so I can use the queued mock responses in the `OrderStore` unit tests. 

### Pull-to-refresh

1. Follow the steps described in #1962. Confirm that you can no longer reproduce the bug. 
2. Perform the same steps in #1962 but on one of the Search → {Filter} pages. The effect should be the same. 

### Paging

1. Navigate to the Processing tab. 
2. Pull to refresh. 
3. Navigate to the Orders tab. Confirm that there are orders preloaded on the list. 
4. Scroll down. Confirm that the paging still works. 
5. Staying on the Orders tab, pull to refresh. 
6. Navigate to the Processing tab.
7. Scroll down. Confirm that the paging still works. 

## Reviewing 

- Focusing on the core changes described in the **Changes** section above would be best.
- You can also follow the [commits progression](https://github.com/woocommerce/woocommerce-ios/pull/2017/commits) which should show how this was all built, piece by piece.
- Please review the unit tests too! 
- Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [x] If it's feasible, I have added unit tests. 
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by shiki** [comment]
@jaclync @pmusolino I handled all the comments and replied to your concerns above. Please let me know what you think. 🙂

**Comment by jkmassel** [comment]
Howdy – we're freezing `3.9` today, so this PR is being bumped to `4.0`. If needs to be released as part of `4.0`, please re-target it at the `release/4.0` branch.

File: Networking/Networking/Network/MockupNetwork.swift [hunk 1]
```diff
@@ -132,8 +132,8 @@ private extension MockupNetwork {
     func filename(for request: URLRequestConvertible) -> String? {
         let searchPath = path(for: request)
         if useResponseQueue {
-            if var queue = responseQueue.filter({ searchPath.hasSuffix($0.key) }).first?.value {
-                return queue.dequeue()
+            if let keyAndQueue = responseQueue.first(where: { searchPath.hasSuffix($0.key) }) {
+                return responseQueue[keyAndQueue.key]?.dequeue()
```
[the remark this answers is not in the document]
Comment (jaclync) [reply Networking/Networking/Network/MockupNetwork.swift]: ohh I missed the explanation in the commit message! this makes sense with the value types

File: Yosemite/Yosemite/Stores/OrderStore.swift [hunk 2]
```diff
@@ -77,6 +83,95 @@ private extension OrderStore {
...
+                    if let orders = orders {
+                        if deleteAllBeforeSaving {
+                            deleteAllOrdersOnce()
+                        }
+
+                        self.upsertStoredOrdersInBackground(readOnlyOrders: orders, onCompletion: completion)
+                    } else if let error = error {
+                        fetchErrors.append(error)
+                        completion()
+                    }
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/OrderStore.swift]: just wondering, is it ever possible the `orders` and `error` are both nil? not sure if it's worth handling the `else` case just in any case?

File: Yosemite/YosemiteTests/Stores/OrderStoreTests+FetchFilteredAndAllOrders.swift [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        store.onAction(action)
+
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    func insert(order: Networking.Order) {
+        let storageOrder = viewStorage.insertNewObject(ofType: Storage.Order.self)
+        storageOrder.update(with: Fixtures.order)
```
Comment (jaclync) [inline Yosemite/YosemiteTests/Stores/OrderStoreTests+FetchFilteredAndAllOrders.swift]: perhaps we can add this helper to [`MockupStorage+Sample.swift`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/YosemiteTests/Mockups/MockupStorage+Sample.swift)?

**Comment by jaclync** [comment]
> @jaclync @pmusolino You bring up a good point though that Orders and Products are synchronized differently. Should we open up a different discussion for this? Probably a p2?

I think it makes sense to have different sync behaviors based on how frequent a tab is updated on average, like the Orders and Reviews tabs are updated more frequently than the Products tab. It would be nice to have it documented for future references for sure, and see how Android handles the syncs as well!

File: WooCommerce/Classes/Tools/SyncCoordinator.swift [hunk 4]
```diff
@@ -9,7 +9,10 @@ protocol SyncingCoordinatorDelegate: class {
     /// The receiver is expected to synchronize the pageNumber. On completion, it should indicate if the sync was
     /// successful or not.
     ///
-    func sync(pageNumber: Int, pageSize: Int, onCompletion: ((Bool) -> Void)?)
+    /// - Parameter reason: A value passed from `resynchronize` or `synchronizeFirstPage`. This can
+    ///                     be used to decide how to perform the sync.
+    ///
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: ((Bool) -> Void)?)
```
[the remark this answers is not in the document]
Comment (shiki) [reply WooCommerce/Classes/Tools/SyncCoordinator.swift]: That's correct. It's "reusable" by different ViewControllers so it has to be a type that any one of them can use.

File: Yosemite/Yosemite/Stores/OrderStore.swift (hunk 2 above)
Comment (shiki) [reply Yosemite/Yosemite/Stores/OrderStore.swift]: I don't think so. I think this should be an `Either<[Order], Error]` or `Result<[Order], Error>`. But we don't have that construct yet. So in here, I'm using `if let` as a shortcut for _optional checking_.

File: Yosemite/YosemiteTests/Stores/OrderStoreTests+FetchFilteredAndAllOrders.swift (hunk 3 above)
Comment (shiki) [reply Yosemite/YosemiteTests/Stores/OrderStoreTests+FetchFilteredAndAllOrders.swift]: Ooohh. Didn't know that that existed. 

Do you mean the `Fixtures.order` or the `insert()` function?

File: Yosemite/Yosemite/Stores/OrderStore.swift (hunk 2 above)
Comment (shiki) [reply Yosemite/Yosemite/Stores/OrderStore.swift]: > not sure if it's worth handling the else case just in any case?

Added thorough `completion` call in f12cb3a010ed1393d620a49bd989ad6a6890d6af. Thanks for catching that!

**Comment by shiki** [comment]
Thanks for the review, @jaclync, and @pmusolino!

**Pull request merged**